### PR TITLE
CI: resolve two issues with dependencies (docs builds and `oldestdeps`)

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
~This is a quick-and-dirty workaround to make CI useful again until the real problem is addressed upstream.~

[example logs](https://github.com/astropy/astropy/actions/runs/10234365497)
[upstream patch](https://github.com/asdf-format/asdf/pull/1815)

edit: since my upstream patch was released very quickly, this is now only targetting oldestdeps, and I do not consider this a temporary patch anymore (at least not until asdf < 3.4.0 is dropped)

edit again: this PR has slightly moved in scope as 2 different situations unfolded at the same time. It's final scope is "get *all* CI back to green". Here are the two problems addressed here:
- `attrs` v24.1.0 deprecated an option that was used in `asdf` (which is an indirect dependency to us). I helped resolve the issue upstream, but we still need to take action here specifically in `oldestdeps` so we don't get a broken install
- `matplotlib` 3.9.1 was yanked (https://github.com/matplotlib/matplotlib/issues/28551), which broke our docs builds

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Windows tests are still down but it doesn't seem related, so I'll undraft this.

**Comment by neutrinoceros** [comment]
I diffed the last-good env from a windows run against this one and here's the result:
```
< astropy @ file:///D:/a/astropy/astropy/.tox/.tmp/package/1/astropy-7.0.0.dev506%2Bgd261cba411.tar.gz#sha256=675995dc7aa23c81b48828469b2722aacdf5e79f480b982294c87c353790cb7a
---
> astropy @ file:///D:/a/astropy/astropy/.tox/.tmp/package/1/astropy-7.0.0.dev500%2Bgc9c2401182.tar.gz#sha256=8d8193cb19e3ba685b2dc42cfbfaca68f261def0640aa37eb2f07d9929d96413
40c40
< hypothesis==6.108.5
---
> hypothesis==6.108.6
92c92
< pyzmq==26.0.3
---
> pyzmq==26.1.0
```
The error we're getting is in matplotlib, which doesn't seem related, so, most likely, this is either a fluke or something changed in the docker image itself. I'd recommend trying to run a second time. Also I might not respond too quickly as I'm on vacation, so if anyone needs to adjust this PR for any reason please go ahead without waiting for my response.

**Comment by neutrinoceros** [comment]
It seems that the remaining issue with windows is due to (or at least, related to) https://github.com/matplotlib/matplotlib/issues/28551
This issue isn't new, and only affects matplotlib 3.9.1 on windows. If we didn't hit it previously maybe it's just a very unlucky draw and maybe we can afford ignoring it for the time being ? (I still recommend re-running to see if it's reproducible here)

**Comment by neutrinoceros** [comment]
Asdf 3.4.0 was just released with the permanent fix so this should not be needed anymore

**Comment by neutrinoceros** [comment]
actually oldestdeps still needs it, so let me adjust the patch

**Comment by tacaswell** [comment]
You probably want to exclude 3.9.1 and/or set the matplotlib dependency to be `--only-binary`

**Comment by neutrinoceros** [comment]
@tacaswell I guess it depends on which timescale we should expect matplotlib to resolve the issue.

**Comment by tacaswell** [comment]
--only-binary on all dependencies with extensions is probably good defensive practice anyway.

Hopefully tonight or tomorrow, but I'm still not sure what the best fix is.

**Comment by tacaswell** [comment]
I yanked the release.  It is a 🔨 but appears to be the only tool available on a Sunday night.

File: tox.ini
Comment (neutrinoceros) [inline tox.ini:119]: note to reviewers: this change is not related to the issue with `attrs`, but emerged from the discussion with mpl maintainer Thomas Caswell; it is intended to protect CI against situations like the one we saw yesterday on https://github.com/matplotlib/matplotlib/issues/28551

**Comment by neutrinoceros** [comment]
@pllim there are no upstream fixes to wait for any longer

File: pyproject.toml
Comment (pllim) [inline pyproject.toml:125]: If 3.9.1 is yanked, we shouldn't do a `==` pin to it? Am I missing something?
```suggestion
    "matplotlib!=3.9.1",  # https://github.com/matplotlib/matplotlib/issues/28234 # TODO: switch to >=3.9.2 when available (3.9.1 is yanked)
```

File: tox.ini
Comment (pllim) [inline tox.ini:80]: Easier to grep this way
```suggestion
    oldestdeps: attrs < 24.1.0  # ASDF_LT_3_4
```

File: tox.ini
Comment (pllim) [reply tox.ini:119]: I also vaguely remember a time where "prefer binary" bit us but I don't remember why, so why not.

File: pyproject.toml
Comment (neutrinoceros) [reply pyproject.toml:125]: Our docs really does require >=3.9.1 at the moment, and because it's yanked but no newer version is available yet, the only way to get it is to require exactly this version.

Comment (pllim) [reply pyproject.toml:125]: Ah, https://github.com/matplotlib/matplotlib/issues/28234 ... Well, that's a headache. Feels wrong to pin to something that is yanked. Let me think a little.

**Comment by neutrinoceros** [comment]
Goodness gracious, macos-intel is down because `bottleneck` 1.4.0 doesn't have wheels for it and my inclusion of `--prefer-binary` leads to a numpy2-incompatible wheel being picked instead. I'll remove this flag.

**Comment by neutrinoceros** [comment]
(good call for adding `Extra CI` here @pllim !)

File: pyproject.toml
Comment (pllim) [reply pyproject.toml:125]: I think matplotlib 3.9.1 still available from conda-forge (https://anaconda.org/conda-forge/matplotlib)? So, theoretically, we could keep this pin untouched and install matplotlib from conda-forge instead for RTD?

https://github.com/astropy/astropy/blob/384e90b5132e386faaf168fdb1bda100d3eb26ef/.readthedocs.yaml#L13-L14

https://github.com/astropy/astropy/blob/384e90b5132e386faaf168fdb1bda100d3eb26ef/docs/rtd_environment.yaml#L5

Comment (pllim) [reply pyproject.toml:125]: Hopefully this situation is short-lived enough upstream that we don't have to worry about users trying to build docs locally?

**Comment by pllim** [comment]
(I knew the prefer binary thing sometimes bites back... 😬 )

File: pyproject.toml
Comment (neutrinoceros) [reply pyproject.toml:125]: I don't feel strongly about it but yes, it seems that the situation will be resolved upstream in a couple days if I understand correctly. Feel free to push to this branch, I'll be AFK for the rest of the day !

Comment (tacaswell) [reply pyproject.toml:125]: It is only yanked because there was no other way to satisfy:

1. do not ship windows wheels that segfault
2. do not cause windows machines that are not set up to compile c/c++ try to install Matplotlib from source 

If this pinning only runs on linux, there is no problem.

**Comment by tacaswell** [comment]
> (I knew the prefer binary thing sometimes bites back... 😬 )

this is one of the sticking points in making pip prefer binary by default.

**Comment by pllim** [comment]
Well, my way to install matplotlib from conda-forge for RTD broke linkcheck. But I think that is probably acceptable if this problem is short-lived. I don't think `==` pinning on a release that is yanked.

**Review by pllim — APPROVED** [review]
Thanks!

**Pull request merged**
